usrsctplib/netinet/sctp_callout.c

```diff
@@ -75,72 +76,201 @@ uint32_t sctp_get_tick_count(void) {
+ *                                  current callout pointer is changed
  */
-static sctp_os_timer_t *sctp_os_timer_next = NULL;
 static sctp_os_timer_t *sctp_os_timer_current = NULL;
-static int sctp_os_timer_waiting = 0;
-static int sctp_os_timer_wait_ctr = 0;
 static userland_thread_id_t sctp_os_timer_current_tid;
+static userland_cond_t sctp_os_timer_current_changed;
 
-/*
- * SCTP_TIMERWAIT_LOCK (sctp_os_timerwait_mtx) protects:
- * - sctp_os_timer_wait_cond: waiting for callout to complete
- * - sctp_os_timer_done_ctr: value of "wait_ctr" after triggering "waiting"
- */
-userland_mutex_t sctp_os_timerwait_mtx;
-static userland_cond_t sctp_os_timer_wait_cond;
-static int sctp_os_timer_done_ctr = 0;
+#if defined(__Userspace__)
+
+static void
+sctp_userland_cond_wait(userland_cond_t* cond,
+                        userland_mutex_t* mtx) {
+	// callee must be robust to spurious wakeups, both because
+	// wakeup could happen due to native API behavior and because
+	// there is indendend timeout for wait.
+#if defined(__Userspace_os_Windows)
+	const DWORD timeoutMillis = 20 * 1000;
+	const BOOL waited = SleepConditionVariableCS(cond, mtx, timeoutMillis);
+	if (!waited) {
+		SCTP_PRINTF("WARN; SleepConditionVariableCS did not return within %ul millis\n", timeoutMillis);
+	}
+#else
+	struct timespec ts;
+	ts.tv_sec = 20;
+	ts.tv_nsec = 0;
+	int rc = pthread_cond_timedwait(cond, mtx, &ts);
+	if (rc) {
+		if (rc == ETIMEDOUT) {
+			SCTP_PRINTF("WARN; pthread_cond_timedwait did not return within %" PRId64 " sec\n",
+				(int64_t)ts.tv_sec);
+		} else {
+			SCTP_PRINTF("ERROR; return code from pthread_cond_wait is %d\n",
+				rc);
+		}
+	}
+#endif
+}
+
+static void
+sctp_userland_cond_signal(userland_cond_t* cond) {
+#if defined(__Userspace_os_Windows)
+	WakeAllConditionVariable(cond);
+#else
+	int rc = pthread_cond_broadcast(cond);
+	if (rc) {
+		SCTP_PRINTF("ERROR; return code from pthread_cond_broadcast is %d\n", rc);
+	}
+#endif
+}
 
+static void
+sctp_userland_cond_init(userland_cond_t* cond) {
+#if defined(__Userspace_os_Windows)
+	InitializeConditionVariable(cond);
+#else
+	int rc = pthread_cond_init(cond, NULL);
+	if (rc) {
+		SCTP_PRINTF("ERROR; return code from pthread_cond_init is %d\n", rc);
+	}
+#endif
+}
+
+static void
+sctp_userland_cond_destroy(userland_cond_t* cond) {
+#if defined(__Userspace_os_Windows)
+	DeleteConditionVariable(cond);
+#else
+	int rc = pthread_cond_destroy(cond);
+	if (rc) {
+		SCTP_PRINTF("ERROR; return code from pthread_cond_destroy is %d\n", rc);
+	}
+#endif
+}
+#endif
+
+static int
+sctp_os_timer_cancel_impl(sctp_os_timer_t* c) {
+	// Assume SCTP_TIMERQ_LOCK done by caller
+
+	c->c_flags &= ~(SCTP_CALLOUT_ACTIVE);
+	c->c_flags |= (SCTP_CALLOUT_CANCELLED);
+
+	if (c->c_flags & SCTP_CALLOUT_PENDING)
+	{
+		c->c_flags &= ~SCTP_CALLOUT_PENDING;
+		sctp_binary_heap_remove(&SCTP_BASE_INFO(timers_queue), &c->heap_node);
+		SCTPDBG(SCTP_DEBUG_TIMER2, "%s: now=%" PRIu32 ": cancelled pending callout %p\n",
```

Comment:
Notes for reviewers:
`SCTPDBG(SCTP_DEBUG_TIMER2, ...)` has proven being invaluable source of information during fix development for debugging with logs.